docs/smtchecker.rst

```diff
@@ -0,0 +1,733 @@
+            require(_a.length >= 5);
+            uint m = 0;
+            for (uint i = 0; i < _a.length; ++i)
+                if (_a[i] > m)
+                    m = _a[i];
+
+            for (uint i = 0; i < _a.length; ++i)
+                assert(m > _a[i]);
+
+            return m;
+        }
+    }
+
+gives us:
+
+.. code-block:: bash
+
+  Warning: CHC: Assertion violation happens here.
+  Counterexample:
+
+  _a = [0, 0, 0, 0, 0]
+   = 0
+
+  Transaction trace:
+  Test.constructor()
+  Test.max([0, 0, 0, 0, 0])
+    --> max.sol:14:4:
+     |
+  14 | 			assert(m > _a[i]);
+
+
+State Properties
+================
+
+So far the examples only demonstrated the use of the SMTChecker over pure code,
+proving properties about specific operations or algorithms.
+A common type of properties in smart contracts are properties that involve the
+state of the contract. Multiple transactions might be needed to make an assertion
+fail for such a property.
+
+As an example, consider a 2D grid where both axis have coordinates in the range (-2^128, 2^128 - 1).
+Let us place a robot at position (0, 0). The robot can only move diagonally, one step at a time,
+and cannot move outside the grid. The robot's state machine can be represented by the smart contract
+below.
+
+.. code-block:: Solidity
+
+    pragma solidity >=0.8.0;
+    pragma experimental SMTChecker;
+    // This may report a warning if no SMT solver available.
+
+
+    contract Robot {
+        int x = 0;
+        int y = 0;
+
+        modifier wall {
+            require(x > type(int128).min && x < type(int128).max);
+            require(y > type(int128).min && y < type(int128).max);
+            _;
+        }
+
+        function moveLeftUp() wall public {
+            --x;
+            ++y;
+        }
+
+        function moveLeftDown() wall public {
+            --x;
+            --y;
+        }
+
+        function moveRightUp() wall public {
+            ++x;
+            ++y;
+        }
+
+        function moveRightDown() wall public {
+            ++x;
+            --y;
+        }
+
+        function inv() public view {
+            assert((x + y) % 2 == 0);
+        }
+    }
+
+Function ``inv`` represents an invariant of the state machine that ``x + y``
+must be even.
+The SMTChecker manages to prove that regardless how many commands we give the
+robot, even if infinitely many, the invariant can *never* fail. The interested
+reader may want to prove that fact manually as well.  Hint: this invariant is
+inductive.
+
+We can also trick the SMTChecker into giving us a path to a certain position we
+think might be reachable.  We can add the property that (2, 4) is *not*
+reachable, by adding the following function.
+
+.. code-block:: Solidity
+
+    function reach_2_4() public view {
+        assert(!(x == 2 && y == 4));
+    }
+
+This property is false, and while proving that the property is false,
+the SMTChecker tells us exactly *how* to reach (2, 4):
```

Comment:
Maybe you want to explain that this trace is not exactly deterministic (it shows *a potential path*). It could show different paths on different machines, different (solidity,z3,chc,..) releases, different resource limits.